execution/src/sce_ledger.rs

```diff
@@ -0,0 +1,336 @@
+                // apply other's updates to cur_entry
+                // TODO avoid clone, act directly on mutable cur_entry
+                let mut res_entry = cur_entry.clone();
+                res_entry.apply_entry_update(other_entry_update);
+                SCELedgerChange::Set(res_entry)
+            }
+
+            // self updates, other updates
+            (
+                SCELedgerChange::Update(cur_entry_update),
+                SCELedgerChange::Update(other_entry_update),
+            ) => {
+                // try to apply other's updates to self's updates
+                // TODO avoid clone, act directly on mutable cur_entry_update
+                let mut res_update = cur_entry_update.clone();
+                res_update.apply_entry_update(other_entry_update);
+                SCELedgerChange::Update(res_update)
+            }
+        };
+        *self = new_val;
+    }
+}
+
+/// SCE ledger
+#[derive(Debug, Clone, Default)]
+pub struct SCELedger(AddressHashMap<SCELedgerEntry>);
+
+/// list of ledger changes (deletions, resets, updates)
+#[derive(Debug, Clone, Default)]
+pub struct SCELedgerChanges(pub AddressHashMap<SCELedgerChange>);
+
+impl SCELedgerChanges {
+    /// extends the current SCELedgerChanges with another
+    pub fn apply_changes(&mut self, changes: &SCELedgerChanges) {
+        for (addr, change) in changes.0.iter() {
+            self.apply_change(*addr, change);
+        }
+    }
+
+    /// appliees a single change to self
+    pub fn apply_change(&mut self, addr: Address, change: &SCELedgerChange) {
+        self.0
+            .entry(addr)
+            .and_modify(|cur_c| cur_c.apply_change(change))
+            .or_insert_with(|| change.clone());
+    }
+}
+
+impl SCELedger {
+    /// applies ledger changes to ledger
+    pub fn apply_changes(&mut self, changes: &SCELedgerChanges) {
+        for (addr, change) in changes.0.iter() {
+            match change {
+                // delete entry
+                SCELedgerChange::Delete => {
+                    self.0.remove(addr);
+                }
+
+                // set entry to absolute value
+                SCELedgerChange::Set(new_entry) => {
+                    self.0.insert(*addr, new_entry.clone());
+                }
+
+                // update entry
+                SCELedgerChange::Update(update) => {
+                    // insert default if absent
+                    self.0
+                        .entry(*addr)
+                        .or_insert_with(|| SCELedgerEntry::default())
+                        .apply_entry_update(update);
+                }
+            }
+        }
+    }
+}
+
+/// represents an execution step from the point of view of the SCE ledger
+/// applying cumulative_history_changes then caused_changes to final_ledger yields the current ledger during the ledger step
+#[derive(Debug, Clone)]
+pub struct SCELedgerStep {
+    // arc/mutex reference to the final ledger
+    pub final_ledger: Arc<Mutex<SCELedger>>,
```

Comment:
To share the final ledger everywhere without copying it, and also because Wasmer needs a Send structure for the context, and we don't want to make the final ledger Copy. There might be a better solution but I couldn't figure it out.

Reply:
Format that as a docment

Reply:
done